### PR TITLE
Enable `decoder_attention_mask` in `generate` function

### PR DESCRIPTION
**Pull request by samuelpullely**
# What does this PR do?

The documentation for `model_kwargs` in the generate function states that model specific keyword arguments will be forwarded to the `forward` function of the model.

https://github.com/huggingface/transformers/blob/4cf38148dc98b3df1df6eb2f06e4f02448026b19/src/transformers/generation/utils.py#L1216-L1219

However, this is currently not the case for the model specific keyword argument `decoder_attention_mask` when it is passed to the `generate` function.

This PR makes the necessary adjustments such that `decoder_attention_mask` can be passed to `generate` and will be used as an optional input in the `forward` function. More precisely, it is now possible to specify `decoder_input_ids` and `decoder_attention_mask` in the `generate` function for `BartForConditionalGeneration` such that some tokens in `decoder_input_ids` are masked.

To illustrate the change, we can use the (slightly modified) mask filling example from https://huggingface.co/docs/transformers/model_doc/bart#mask-filling

```python
import torch
from transformers import BartForConditionalGeneration, BartTokenizer

model = BartForConditionalGeneration.from_pretrained("facebook/bart-large", forced_bos_token_id=0)
tokenizer = BartTokenizer.from_pretrained("facebook/bart-large")
sentence = "UN Chief Says There Is No <mask> in Syria"
batch = tokenizer(sentence, return_tensors="pt")
padding_size = 3
decoder_input_ids = torch.tensor(
    [[model.config.decoder_start_token_id] + padding_size * [model.config.pad_token_id] + [model.config.bos_token_id]], 
    dtype=torch.long
)
decoder_attention_mask = torch.where(decoder_input_ids==model.config.pad_token_id, 0, 1)
generated_ids = model.generate(input_ids=batch["input_ids"], use_cache=False, max_new_tokens=20, 
                               decoder_input_ids=decoder_input_ids, decoder_attention_mask=decoder_attention_mask)
decoded = tokenizer.batch_decode(generated_ids)
for seq in decoded:
    print(seq)
```

Note that `use_cache=False` is required when `decoder_input_ids` is manually specified.

Output without PR:

```
</s><pad><pad><pad><s><s>,</s>
```
Clearly, padding tokens are not ignored because `decoder_attention_mask` is not forwarded to the `forward` function. Hence, the strange output.

Output with PR:

```
</s><pad><pad><pad><s>UN Chief Says There Is No Plan B for Peace in Syria</s>
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.

https://discuss.huggingface.co/t/is-t5-expected-to-ignore-padding-tokens-in-decoder-input-ids-when-decoder-attention-mask-is-not-provided/10271

- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker and @younesbelkada

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Hey! Make sure to run `make fixup` to pass all the quality tests 😉

**Review by younesbelkada — COMMENTED** [review]
Thanks so much for the fix, this seems correct to me
I can also confirm slow tests are passing with this fix (single + multi-gpu tests)
leaving it to @ArthurZucker(make sure to run styling changes as suggested above) !

**Comment by ArthurZucker** [comment]
Style is okay now, but it seems like the `repo-consistency` is not. Running `make repo-consistency` should solve this

**Review by younesbelkada — APPROVED** [review]
Thanks for the fix!
I am ok with this PR with the principle that a user could in practice provide a `decoder_input_ids` with padding tokens (even if it is not that common). 
Let's wait for @ArthurZucker & @sgugger to decide whether this solution is ok to merge

**Review by ArthurZucker — APPROVED** [review]
Thanks it is very clean 😉

**Review by sgugger — COMMENTED** [review]
I'd also like for @gante to have a look here.

**Comment by gante** [comment]
Hi @samuelpullely 👋 

Thank you for the PR and the use-case example with self-debiasing! The PR looks good to me, but it's missing one small detail: tests 🤗  If possible, I'd like to ask you to add a regression test, perhaps in the [Bart integration tests section](https://github.com/huggingface/transformers/blob/76d02feadbc99bbccd86e67b02728338a2469f22/tests/models/bart/test_modeling_bart.py#L841). It can be a copy of your example above -- it would go a long way ensuring we don't regress. Let me know if you need a hand.

**Review by gante — APPROVED** [review]
LGTM 👍

**Comment by gante** [comment]
@sgugger can we merge this PR? :)
